CoreCLR initialization failures logging

janvorli opened the pull request:
This change adds detecting and logging of failures during coreclr initialization. For logging, it uses a new host API `coreclr_set_error_writer` to register a callback to report the errors to the host. The hosts have support for optional usage of this API so that they can work with older runtime versions as well.

The logging checks and reports failures with:
* System.Private.CoreLib.dll
* GC initialization
* JIT initialization
* libSystem.Native.so/dylib on Unix

The logging messages should allow customers to self-diagnose the issues causing the failures.

janvorli committed:
CoreCLR initialization failures logging

This change adds detecting and logging of failures during coreclr
initialization. For logging, it uses a new host API
`coreclr_set_error_writer` to register a callback to report the errors
to the host. The hosts have support for optional usage of this API so
that they can work with older runtime versions as well.

The logging checks and reports failures with:
* System.Private.CoreLib.dll
* GC initialization
* JIT initialization
* libSystem.Native.so/dylib on Unix

The logging messages should allow customers to self-diagnose the issues
causing the failures.

janvorli commented:
Here are some examples of the error logs:
### Missing SPC file
Failed to load System.Private.CoreLib.dll (error code 0x80070057)
Path: 'F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll'

Could not load file or assembly 'F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll'. The parameter is incorrect. (0x80070057 (E_INVALIDARG))

### Another assembly instead of SPC
Failed to load System.Private.CoreLib.dll (error code 0x80131522)
Path: 'F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll'

Could not load type 'System.Object' from assembly 'System.ObjectModel, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.

### Text file instead of SPC
Failed to load System.Private.CoreLib.dll (error code 0x80131018)
Path: 'F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll'

Could not load file or assembly 'F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll'. The module was expected to contain an assembly manifest. (0x80131018 (COR_E_ASSEMBLYEXPECTED))

### Wrong version of SPC (x86)
Failed to load System.Private.CoreLib.dll (error code 0x8007000B)
Path: 'F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll'

Could not load file or assembly 'F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll'. An attempt was made to load a program with an incorrect format. (0x8007000B (COR_E_BADIMAGEFORMAT))

### Wrong version of SPC (x64 from 3.1):
Failed to load System.Private.CoreLib.dll (error code 0x80131522)
Path: 'F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll'

Could not load type 'System.Runtime.CompilerServices.CastHelpers' from assembly 'System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e'.

### OOM
Failed to load System.Private.CoreLib.dll (error code 0x8007000E)
Path:
F:\git\runtime8\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\System.Private.CoreLib.dll
Error message:
Out Of Memory

### Wrong setting of GC log file
GC: Cannot create log file
GC heap initialization failed with error 0x80004005

jkotas commented on src/coreclr/vm/ceemain.cpp:
```diff
@@ -947,6 +950,21 @@ void EEStartupHelper()
 #endif // FEATURE_MINIMETADATA_IN_TRIAGEDUMPS
 
+#ifdef TARGET_UNIX
+        // Preload the libSystem.Native.so/dylib to detect possible problems with loading it early
```
Is this going to work for singlefilehost?

janvorli replied:
Ah, singlefilehost should have been excluded here

AaronRobinsonMSFT commented on src/coreclr/hosts/corerun/corerun.cpp:
```diff
@@ -147,11 +147,11 @@ static string_t build_tpa(const string_t& core_root, const string_t& core_librar
 }
 
-static bool try_get_export(pal::mod_t mod, const char* symbol, void** fptr)
+static bool try_get_export(pal::mod_t mod, const char* symbol, void** fptr, bool isOptional = false)
```
We're not checking the return value for the case where this is optional. I don't think this is needed if we aren't even checking the result. I'd prefer to just ignore the case at the callsite instead of making an optional param that just confuses the usage. The error message below can stay because it is appropriate and makes sense to report that regardless of intent since this is corerun and for testing.

janvorli replied:
Ok, I was thinking that the error message might confuse people, but I guess it is fine to keep it. The optional argument was added just to suppress it.

AaronRobinsonMSFT commented on src/coreclr/hosts/corerun/corerun.cpp:
```diff
@@ -290,6 +296,9 @@ static int run(const configuration& config)
     }
 
+    // The coreclr_set_error_writer is optional
+    try_get_export(coreclr_mod, "coreclr_set_error_writer", (void**)&coreclr_set_error_writer_func, true /* isOptional */);
```
```suggestion
    (void)try_get_export(coreclr_mod, "coreclr_set_error_writer", (void**)&coreclr_set_error_writer_func, true /* isOptional */);
```

AaronRobinsonMSFT commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2049,9 +2050,13 @@ static void LoadAndInitializeJIT(LPCWSTR pwzJitName DEBUGARG(LPCWSTR pwzJitPath)
         else
         {
             LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: invalid characters in %S\n", pwzJitName));
+            MAKE_UTF8PTR_FROMWIDE_NOTHROW(utf8JitName, pwzJitName);
+            LogErrorToHost("LoadAndInitializeJIT: invalid characters in %s", utf8JitName);
```
```suggestion
            MAKE_UTF8PTR_FROMWIDE_NOTHROW(utf8JitName, pwzJitName);
            LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: invalid characters in %s\n", utf8JitName));
            LogErrorToHost("LoadAndInitializeJIT: invalid characters in %s", utf8JitName);
```

AaronRobinsonMSFT commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2105,28 +2110,33 @@ static void LoadAndInitializeJIT(LPCWSTR pwzJitName DEBUGARG(LPCWSTR pwzJitPath)
                     {
                         // Mismatched version ID. Fail the load.
                         LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: mismatched JIT version identifier in %S\n", pwzJitName));
```
```suggestion
                        LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: mismatched JIT version identifier in %s\n", utf8JitName));
```

jkotas commented:
> Missing SPC file

Where is the invalid arg comming from in this case? Would it be feasible to change it to file not found?

AaronRobinsonMSFT commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2105,28 +2110,33 @@ static void LoadAndInitializeJIT(LPCWSTR pwzJitName DEBUGARG(LPCWSTR pwzJitPath)
                 }
                 else
                 {
                     LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: failed to get ICorJitCompiler in %S\n", pwzJitName));
```
```suggestion
                    LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: failed to get ICorJitCompiler in %s\n", utf8JitName));
```

AaronRobinsonMSFT commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2105,28 +2110,33 @@ static void LoadAndInitializeJIT(LPCWSTR pwzJitName DEBUGARG(LPCWSTR pwzJitPath)
             }
             else
             {
                 LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: failed to find 'getJit' entrypoint in %S\n", pwzJitName));
```
```suggestion
                LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: failed to find 'getJit' entrypoint in %s\n", utf8JitName));
```

AaronRobinsonMSFT commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2105,28 +2110,33 @@ static void LoadAndInitializeJIT(LPCWSTR pwzJitName DEBUGARG(LPCWSTR pwzJitPath)
         }
         EX_CATCH
         {
             LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: caught an exception trying to initialize %S\n", pwzJitName));
```
```suggestion
            LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: caught an exception trying to initialize %s\n", utf8JitName));
```

AaronRobinsonMSFT commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2105,28 +2110,33 @@ static void LoadAndInitializeJIT(LPCWSTR pwzJitName DEBUGARG(LPCWSTR pwzJitPath)
     else
     {
         pJitLoadData->jld_hr = hr;
         LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: failed to load %S, hr=0x%08x\n", pwzJitName, hr));
```
```suggestion
        LOG((LF_JIT, LL_FATALERROR, "LoadAndInitializeJIT: failed to load %s, hr=0x%08x\n", utf8JitName, hr));
```

AaronRobinsonMSFT reviewed (commented):
I like the work here a lot.

janvorli commented:
> Where is the invalid arg coming from in this case? Would it be feasible to change it to file not found?

It comes from here:
https://github.com/dotnet/runtime/blob/89cfbfd84146e89ab18771c42928d23ff09232c1/src/coreclr/binder/utils.cpp#L202

When the S.P.C. is not found at the expected location, corerun tries to find it in the TPA list. To iterate over the list, it calls the `GetNextTPAPath` with `dllOnly` argument set to `true`. The `E_INVALIDARG` is returned because it comes to the `corerun.exe` on the TPA list. The behavior of the iteration is kind of weird, as it stops iterating over the list. I would expect it to skip non-dll entries instead in this case.
The iteration loop is here:
https://github.com/dotnet/runtime/blob/89cfbfd84146e89ab18771c42928d23ff09232c1/src/coreclr/binder/assemblybindercommon.cpp#L297-L315

janvorli committed:
Reflect PR feedback and a bit of cleanup

* revert the try_get_export to not to have the isOptional argument
* Unify the error logging in JIT loading to single variadic macro
* Also fixes Unix build, the code in gcenv.unix.cpp cannot use the GCToEEInterface

jkotas commented:
> I would expect it to skip non-dll entries instead in this case.

Yes, I agree. (A separate question is why we need to be looking for CoreLib in so many places.)

janvorli committed:
Add single file host detection

jkotas commented on src/coreclr/vm/ceemain.cpp:
```diff
@@ -947,6 +950,29 @@ void EEStartupHelper()
 #endif // FEATURE_MINIMETADATA_IN_TRIAGEDUMPS
 
+#ifdef TARGET_UNIX
+        // Check if the current code is executing in the single file host or in libcoreclr.so. The libSystem.Native is linked
```
Would it be better to just check `if (!g_coreclr_embedded)`?

janvorli replied:
I was looking for a check like that and have failed to locate this one :-)

elinor-fung replied:
I think you could also use the `CORECLR_EMBEDDED` define?

janvorli replied:
Ah, nice. I have not noticed that the ceemain.cpp is one of the few files that are compiled separately for single file host and libcoreclr.so.

janvorli replied:
@elinor-fung I have found that when S.P.C. is bundled, this `NativeLibrary::LoadLibraryByName` call doesn't work as the S.P.C. doesn't exist in the filesystem, just in the bundle. This causes e.g. the 
 Microsoft.NET.HostModel.Tests.BundleAndRun.TestWithRelativePaths` test to fail.
Do I need to pass something else than `SystemDomain::SystemAssembly()` to `NativeLibrary::LoadLibraryByName` to make it work?

janvorli replied:
Ah, it seems I should pass in the DLLIMPORTSEARCHPATH_ASSEMBLYDIRECTORY.

janvorli replied:
The problem was actually different, in the `EEStartupHelper`, the native dll search paths from the properties passed by the host are not set yet. Moving the check to the end of the `CreateAppDomainWithManager` resolved the issue.

janvorli replied:
Sigh, yet one more host test failing.

elinor-fung replied:
I think that might be a variation of https://github.com/dotnet/runtime/issues/75387? The case that is failing uses an extraction path that has a backslash on Unix and loading native libraries in `PAL_LoadLibraryDirect` goes through `FILEDosToUnixPathA`.

janvorli replied:
Ah, right, running it locally with the `FILEDosToUnixPathA` modified to be dummy fixed the issue. There was one more mono related issue that I've fixed, but now it seems fine.

janvorli committed:
Fix libSystem.Native check location

janvorli committed:
Add the coreclr_set_error_writer API to Mono

janvorli committed:
Fix Mono x86 missing calling convention spec

janvorli commented:
The `AppHost.Bundle.Tests.BundleExtractToSpecificPath.Bundle_Extraction_To_Relative_Path_Succeeds` fails with these changes for this `[InlineData("foo\\bar", BundleOptions.BundleNativeBinaries)]`. This test case should be Windows specific after this change as it expects the `foo` and `bar` to be separate directories. We already have an exclusion check for similar case with forward slash on Windows, so I will mimic that.

janvorli commented:
@jkotas, @AaronRobinsonMSFT the CI tests have passed except for one known issue and one unrelated issue. I believe I've addressed your feedback. Can you please take another look?

jkotas commented:
Could you please make the message for missing CoreLib better (file not found instead of invalid arg)?

janvorli committed:
Fix error code returned when the S.P.C. is not found

janvorli commented:
@jkotas I've fixed that in the last commit.

jkotas reviewed (approved):
LGTM. Thank you!

vitek-karas commented on src/coreclr/dlls/mscoree/exports.cpp:
```diff
@@ -158,6 +158,26 @@ static void ConvertConfigPropertiesToUnicode(
+//
+// Set callback for writing error logging
+//
+// Parameters:
+//  errorWriter             - callback that will be called for each line of the error info
+//                          - passing in NULL removes a callback that was previously set
+//
```
We should probably add that the callback can be called from any thread and thus must be thread safe.

janvorli replied:
It is not the case at the moment, it is always called from the thread running the initialization. Do you mean that we should add that comment to make it possible if we ever needed that?

vitek-karas replied:
Didn't realize we don't need it currently. But I would add it anyway... it's really easy to forget about this and it could break in really weird ways (the current host implementation is thread safe fortunately).

janvorli commented:
The AllSubsets_Mono_LLVMFullAot_RuntimeTests llvmfullaot legs for x64 and arm64 are failing due to a sigsegv after I've pushed the fix for the error code. I don't see how that could have introduced such issue and I am unable to repro it on my local Ubuntu 21.04. Trying to repro in the centos 7 container that the lab uses.

MichalStrehovsky commented:
> The AllSubsets_Mono_LLVMFullAot_RuntimeTests llvmfullaot legs for x64 and arm64 are failing due to a sigsegv after I've pushed the fix for the error code. I don't see how that could have introduced such issue and I am unable to repro it on my local Ubuntu 21.04. Trying to repro in the centos 7 container that the lab uses.

I think you're hitting https://github.com/dotnet/runtime/pull/78580#issuecomment-1340069275 and it's not your PR.

janvorli commented:
@MichalStrehovsky ah, great, thank you for letting me know!
